From a6be616109d341445b10827ee471421792154700 Mon Sep 17 00:00:00 2001
From: Etienne Carriere <etienne.carriere@foss.st.com>
Date: Tue, 22 Oct 2024 18:07:46 +0200
Subject: [PATCH 1460/1695] plat-stm32mp1: scmi_server: test clock access
 against ETZPC config

Check whether or not an SCMI clock can be accessed by non-secure client
based on the ETZPC DECPROT configuration instead of relying on
shared_resources.c driver. This latter is now useless since integration
of the firewall framework and will be soon removed.

On STM32MP15 based platforms, this impacts SCMI clock and reset domains
related to CRPY1, HASH1, I2C4, I2C6, IWDG1, RNG1, SPI6 and USART1. On
these platforms, regarding the GPIOZ resources, the GPIO bank SCMI clock
is always accessible but the GPIOZ SCMI reset domain is only accessible
when the full GPIOZ bank is assign to non-secure, because their maybe
be individual GPIOZ pins assigned to non-secure and some assigned to
secure world.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/optee_os.git 42ca717bcbc3663d54f58fee4c5cecab35a0a9c4 plat-stm32mp1: scmi_server: test clock access against ETZPC config]
Change-Id: I80c6fb04c03dfac1e3fc048d948bd6e1b687f690
Link: https://github.com/OP-TEE/optee_os/commit/b18ace9b2cebb491aa3a2b7544a2f16b9a0b41bb
Signed-off-by: Etienne Carriere <etienne.carriere@foss.st.com>
Reviewed-on: https://gerrit.st.com/c/mpu/oe/optee/optee_os/+/413522
Domain-Review: Yann GAUTIER <yann.gautier@foss.st.com>
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
Reviewed-by: Thomas BOURGOIN <thomas.bourgoin@foss.st.com>
---
 core/arch/arm/plat-stm32mp1/scmi_server.c | 166 ++++++++++++++++------
 1 file changed, 119 insertions(+), 47 deletions(-)

diff --git a/core/arch/arm/plat-stm32mp1/scmi_server.c b/core/arch/arm/plat-stm32mp1/scmi_server.c
index 9ae82d810..c6c01eedd 100644
--- a/core/arch/arm/plat-stm32mp1/scmi_server.c
+++ b/core/arch/arm/plat-stm32mp1/scmi_server.c
@@ -8,17 +8,19 @@
 #include <confine_array_index.h>
 #include <drivers/clk.h>
 #include <drivers/clk_dt.h>
+#include <drivers/firewall_device.h>
 #include <drivers/regulator.h>
 #include <drivers/rstctrl.h>
 #include <drivers/scmi-msg.h>
 #include <drivers/scmi.h>
 #include <drivers/stm32_cpu_opp.h>
-#include <drivers/stm32mp_dt_bindings.h>
 #include <drivers/stm32_etzpc.h>
 #include <drivers/stm32_remoteproc.h>
+#include <drivers/stm32mp_dt_bindings.h>
 #include <initcall.h>
 #include <keep.h>
 #include <kernel/pm.h>
+#include <libfdt.h>
 #include <mm/core_memprot.h>
 #include <mm/core_mmu.h>
 #include <platform_config.h>
@@ -35,15 +37,20 @@
 #define SCMI_RD_NAME_SIZE	16
 #define SCMI_VOLTD_NAME_SIZE	16
 
+#define ETZPC_ID_ALWAYS_ACCESSIBLE	(STM32MP1_ETZPC_MAX_ID + 1)
+#define ETZPC_ID_NEVER_ACCESSIBLE	(STM32MP1_ETZPC_MAX_ID + 2)
+
 /*
  * struct stm32_scmi_clk - Data for the exposed clock
  * @clock_id: Clock identifier in RCC clock driver
+ * @etzpc_id: ETZPC ID of the peripheral related to the clock
  * @name: Clock string ID exposed to channel
  * @enabled: State of the SCMI clock
  */
 struct stm32_scmi_clk {
 	unsigned long clock_id;
 	struct clk *clk;
+	unsigned int etzpc_id;
 	const char *name;
 	bool enabled;
 };
@@ -51,13 +58,13 @@ struct stm32_scmi_clk {
 /*
  * struct stm32_scmi_rd - Data for the exposed reset controller
  * @reset_id: Reset identifier in RCC reset driver
- * @fw_id: Firewall ID of the peripheral
+ * @etzpc_id: ETZPC ID of the peripheral related to the reset controller
  * @name: Reset string ID exposed to channel
  * @rstctrl: Reset controller device
  */
 struct stm32_scmi_rd {
 	unsigned long reset_id;
-	unsigned int fw_id;
+	unsigned int etzpc_id;
 	const char *name;
 	struct rstctrl *rstctrl;
 };
@@ -83,17 +90,37 @@ register_phys_mem(MEM_AREA_IO_NSEC, CFG_STM32MP1_SCMI_SHM_BASE,
 #endif
 #endif /*CFG_STM32MP1_SCMI_SHM_BASE*/
 
+/* SCMI clock always accessible */
 #define CLOCK_CELL(_scmi_id, _id, _name, _init_enabled) \
 	[(_scmi_id)] = { \
 		.clock_id = (_id), \
+		.etzpc_id = ETZPC_ID_ALWAYS_ACCESSIBLE, \
 		.name = (_name), \
 		.enabled = (_init_enabled), \
 	}
 
-#define RESET_CELL(_scmi_id, _id, _fw_id, _name) \
+/* SCMI clock accessible upon DECPROT ID assigned to non-secure */
+#define CLOCK_CELL_DECPROT(_scmi_id, _id, _name, _init_enabled, _etzpc_id) \
+	[(_scmi_id)] = { \
+		.clock_id = (_id), \
+		.etzpc_id = (_etzpc_id), \
+		.name = (_name), \
+		.enabled = (_init_enabled), \
+	}
+
+/* SCMI reset domain always accessible */
+#define RESET_CELL(_scmi_id, _id, _name) \
 	[(_scmi_id)] = { \
 		.reset_id = (_id), \
-		.fw_id = (_fw_id), \
+		.etzpc_id = ETZPC_ID_ALWAYS_ACCESSIBLE, \
+		.name = (_name), \
+	}
+
+/* SCMI reset domain accessible upon DECPROT ID assigned to non-secure */
+#define RESET_CELL_DECPROT(_scmi_id, _id, _name, _etzpc_id) \
+	[(_scmi_id)] = { \
+		.reset_id = (_id), \
+		.etzpc_id = (_etzpc_id), \
 		.name = (_name), \
 	}
 
@@ -144,43 +171,59 @@ static struct stm32_scmi_clk stm32_scmi_clock[] = {
 	CLOCK_CELL(CK_SCMI_MPU, CK_MPU, "ck_mpu", true),
 	CLOCK_CELL(CK_SCMI_AXI, CK_AXI, "ck_axi", true),
 	CLOCK_CELL(CK_SCMI_BSEC, BSEC, "bsec", true),
-	CLOCK_CELL(CK_SCMI_CRYP1, CRYP1, "cryp1", false),
+	CLOCK_CELL_DECPROT(CK_SCMI_CRYP1, CRYP1, "cryp1", false,
+			   STM32MP1_ETZPC_CRYP1_ID),
 	CLOCK_CELL(CK_SCMI_GPIOZ, GPIOZ, "gpioz", false),
-	CLOCK_CELL(CK_SCMI_HASH1, HASH1, "hash1", false),
-	CLOCK_CELL(CK_SCMI_I2C4, I2C4_K, "i2c4_k", false),
-	CLOCK_CELL(CK_SCMI_I2C6, I2C6_K, "i2c6_k", false),
-	CLOCK_CELL(CK_SCMI_IWDG1, IWDG1, "iwdg1", false),
-	CLOCK_CELL(CK_SCMI_RNG1, RNG1_K, "rng1_k", true),
+	CLOCK_CELL_DECPROT(CK_SCMI_HASH1, HASH1, "hash1", false,
+			   STM32MP1_ETZPC_HASH1_ID),
+	CLOCK_CELL_DECPROT(CK_SCMI_I2C4, I2C4_K, "i2c4_k", false,
+			   STM32MP1_ETZPC_I2C4_ID),
+	CLOCK_CELL_DECPROT(CK_SCMI_I2C6, I2C6_K, "i2c6_k", false,
+			   STM32MP1_ETZPC_I2C6_ID),
+	CLOCK_CELL_DECPROT(CK_SCMI_IWDG1, IWDG1, "iwdg1", false,
+			   STM32MP1_ETZPC_IWDG1_ID),
+	CLOCK_CELL_DECPROT(CK_SCMI_RNG1, RNG1_K, "rng1_k", true,
+			   STM32MP1_ETZPC_RNG1_ID),
 	CLOCK_CELL(CK_SCMI_RTC, RTC, "ck_rtc", true),
 	CLOCK_CELL(CK_SCMI_RTCAPB, RTCAPB, "rtcapb", true),
-	CLOCK_CELL(CK_SCMI_SPI6, SPI6_K, "spi6_k", false),
-	CLOCK_CELL(CK_SCMI_USART1, USART1_K, "usart1_k", false),
+	CLOCK_CELL_DECPROT(CK_SCMI_SPI6, SPI6_K, "spi6_k", false,
+			   STM32MP1_ETZPC_SPI6_ID),
+	CLOCK_CELL_DECPROT(CK_SCMI_USART1, USART1_K, "usart1_k", false,
+			   STM32MP1_ETZPC_USART1_ID),
 };
 #endif
 
 #ifdef CFG_STM32MP13
 static struct stm32_scmi_rd stm32_scmi_reset_domain[] = {
-	RESET_CELL(RST_SCMI_LTDC, LTDC_R, STM32MP1_ETZPC_LTDC_ID, "ltdc"),
-	RESET_CELL(RST_SCMI_MDMA, MDMA_R, STM32MP1_ETZPC_MAX_ID, "mdma"),
+	RESET_CELL_DECPROT(RST_SCMI_LTDC, LTDC_R, "ltdc",
+			   STM32MP1_ETZPC_LTDC_ID),
+	RESET_CELL(RST_SCMI_MDMA, MDMA_R, "mdma"),
 };
 #endif
 
 #ifdef CFG_STM32MP15
 static struct stm32_scmi_rd stm32_scmi_reset_domain[] = {
-	RESET_CELL(RST_SCMI_SPI6, SPI6_R, STM32MP1_ETZPC_SPI6_ID, "spi6"),
-	RESET_CELL(RST_SCMI_I2C4, I2C4_R, STM32MP1_ETZPC_I2C4_ID, "i2c4"),
-	RESET_CELL(RST_SCMI_I2C6, I2C6_R, STM32MP1_ETZPC_I2C6_ID, "i2c6"),
-	RESET_CELL(RST_SCMI_USART1, USART1_R, STM32MP1_ETZPC_USART1_ID,
-		   "usart1"),
-	RESET_CELL(RST_SCMI_STGEN, STGEN_R, STM32MP1_ETZPC_STGENC_ID, "stgen"),
-	RESET_CELL(RST_SCMI_GPIOZ, GPIOZ_R, STM32MP1_ETZPC_GPIOZ_ID, "gpioz"),
-	RESET_CELL(RST_SCMI_CRYP1, CRYP1_R, STM32MP1_ETZPC_CRYP1_ID, "cryp1"),
-	RESET_CELL(RST_SCMI_HASH1, HASH1_R, STM32MP1_ETZPC_HASH1_ID, "hash1"),
-	RESET_CELL(RST_SCMI_RNG1, RNG1_R, STM32MP1_ETZPC_RNG1_ID, "rng1"),
-	RESET_CELL(RST_SCMI_MDMA, MDMA_R, STM32MP1_ETZPC_MAX_ID, "mdma"),
-	RESET_CELL(RST_SCMI_MCU, MCU_R, STM32MP1_ETZPC_MAX_ID, "mcu"),
-	RESET_CELL(RST_SCMI_MCU_HOLD_BOOT, MCU_HOLD_BOOT_R,
-		   STM32MP1_ETZPC_MAX_ID, "mcu_hold_boot"),
+	RESET_CELL_DECPROT(RST_SCMI_SPI6, SPI6_R, "spi6",
+			   STM32MP1_ETZPC_SPI6_ID),
+	RESET_CELL_DECPROT(RST_SCMI_I2C4, I2C4_R, "i2c4",
+			   STM32MP1_ETZPC_I2C4_ID),
+	RESET_CELL_DECPROT(RST_SCMI_I2C6, I2C6_R, "i2c6",
+			   STM32MP1_ETZPC_I2C6_ID),
+	RESET_CELL_DECPROT(RST_SCMI_USART1, USART1_R, "usart1",
+			   STM32MP1_ETZPC_USART1_ID),
+	RESET_CELL_DECPROT(RST_SCMI_STGEN, STGEN_R, "stgen",
+			   STM32MP1_ETZPC_STGENC_ID),
+	RESET_CELL_DECPROT(RST_SCMI_GPIOZ, GPIOZ_R, "gpioz",
+			   STM32MP1_ETZPC_GPIOZ_ID),
+	RESET_CELL_DECPROT(RST_SCMI_CRYP1, CRYP1_R, "cryp1",
+			   STM32MP1_ETZPC_CRYP1_ID),
+	RESET_CELL_DECPROT(RST_SCMI_HASH1, HASH1_R, "hash1",
+			   STM32MP1_ETZPC_HASH1_ID),
+	RESET_CELL_DECPROT(RST_SCMI_RNG1, RNG1_R, "rng1",
+			   STM32MP1_ETZPC_RNG1_ID),
+	RESET_CELL(RST_SCMI_MDMA, MDMA_R, "mdma"),
+	RESET_CELL(RST_SCMI_MCU, MCU_R, "mcu"),
+	RESET_CELL(RST_SCMI_MCU_HOLD_BOOT, MCU_HOLD_BOOT_R, "mcu_hold_boot"),
 };
 #endif
 
@@ -321,6 +364,41 @@ const uint8_t *plat_scmi_protocol_list(unsigned int channel_id __unused)
 	return plat_protocol_list;
 }
 
+static bool nsec_can_access_resource(unsigned int etzpc_id)
+{
+	static struct firewall_controller *etzpc_fw_ctrl;
+	uint32_t query_arg = DECPROT(etzpc_id, DECPROT_NS_RW, DECPROT_UNLOCK);
+	struct firewall_query query = {
+		.arg_count = 1,
+		.args = &query_arg,
+		.ctrl = etzpc_fw_ctrl,
+	};
+
+	if (etzpc_id == ETZPC_ID_NEVER_ACCESSIBLE)
+		return false;
+	if (etzpc_id == ETZPC_ID_ALWAYS_ACCESSIBLE)
+		return true;
+
+	if (!etzpc_fw_ctrl) {
+		struct dt_driver_provider *prov = NULL;
+		int node = 0;
+
+		node = fdt_node_offset_by_compatible(get_embedded_dt(), -1,
+						     "st,stm32-etzpc");
+		if (node < 0)
+			panic();
+
+		prov = dt_driver_get_provider_by_node(node, DT_DRIVER_FIREWALL);
+		if (!prov)
+			panic();
+
+		etzpc_fw_ctrl = dt_driver_provider_priv_data(prov);
+		query.ctrl = etzpc_fw_ctrl;
+	}
+
+	return firewall_check_access(&query) == TEE_SUCCESS;
+}
+
 /*
  * Platform SCMI clocks
  */
@@ -350,7 +428,7 @@ const char *plat_scmi_clock_get_name(unsigned int channel_id,
 {
 	struct stm32_scmi_clk *clock = find_clock(channel_id, scmi_id);
 
-	if (!clock || !stm32mp_nsec_can_access_clock(clock->clock_id))
+	if (!clock || !nsec_can_access_resource(clock->etzpc_id))
 		return NULL;
 
 	return clock->name;
@@ -372,7 +450,7 @@ int32_t plat_scmi_clock_rates_array(unsigned int channel_id,
 	if (!clock)
 		return SCMI_NOT_FOUND;
 
-	if (!stm32mp_nsec_can_access_clock(clock->clock_id))
+	if (!nsec_can_access_resource(clock->etzpc_id))
 		return SCMI_DENIED;
 
 	switch (scmi_id) {
@@ -409,7 +487,7 @@ int32_t plat_scmi_clock_rates_by_step(unsigned int channel_id,
 	if (!clock)
 		return SCMI_NOT_FOUND;
 
-	if (!stm32mp_nsec_can_access_clock(clock->clock_id))
+	if (!nsec_can_access_resource(clock->etzpc_id))
 		return SCMI_DENIED;
 
 	switch (scmi_id) {
@@ -435,7 +513,7 @@ int32_t plat_scmi_clock_set_rate(unsigned int channel_id, unsigned int scmi_id,
 	if (!clock)
 		return SCMI_NOT_FOUND;
 
-	if (!stm32mp_nsec_can_access_clock(clock->clock_id))
+	if (!nsec_can_access_resource(clock->etzpc_id))
 		return SCMI_DENIED;
 
 	switch (scmi_id) {
@@ -465,7 +543,7 @@ unsigned long plat_scmi_clock_get_rate(unsigned int channel_id,
 {
 	struct stm32_scmi_clk *clock = find_clock(channel_id, scmi_id);
 
-	if (!clock || !stm32mp_nsec_can_access_clock(clock->clock_id))
+	if (!clock || !nsec_can_access_resource(clock->etzpc_id))
 		return 0;
 
 	return clk_get_rate(clock->clk);
@@ -477,7 +555,7 @@ unsigned long plat_scmi_clock_round_rate(unsigned int channel_id,
 {
 	struct stm32_scmi_clk *clock = find_clock(channel_id, scmi_id);
 
-	if (!clock || !stm32mp_nsec_can_access_clock(clock->clock_id))
+	if (!clock || !nsec_can_access_resource(clock->etzpc_id))
 		return 0;
 
 	return clk_round_rate(stm32mp_rcc_clock_id_to_clk(clock->clock_id), rate);
@@ -487,7 +565,7 @@ int32_t plat_scmi_clock_get_state(unsigned int channel_id, unsigned int scmi_id)
 {
 	struct stm32_scmi_clk *clock = find_clock(channel_id, scmi_id);
 
-	if (!clock || !stm32mp_nsec_can_access_clock(clock->clock_id))
+	if (!clock || !nsec_can_access_resource(clock->etzpc_id))
 		return 0;
 
 	return (int32_t)clock->enabled;
@@ -501,7 +579,7 @@ int32_t plat_scmi_clock_set_state(unsigned int channel_id, unsigned int scmi_id,
 	if (!clock)
 		return SCMI_NOT_FOUND;
 
-	if (!stm32mp_nsec_can_access_clock(clock->clock_id))
+	if (!nsec_can_access_resource(clock->etzpc_id))
 		return SCMI_DENIED;
 
 	if (enable_not_disable) {
@@ -528,7 +606,7 @@ int32_t plat_scmi_clock_get_duty_cycle(unsigned int channel_id, unsigned int scm
 	struct clk *clk_ref = NULL;
 	int ret = 0;
 
-	if (!clock || !stm32mp_nsec_can_access_clock(clock->clock_id))
+	if (!clock || !nsec_can_access_resource(clock->etzpc_id))
 		return SCMI_DENIED;
 
 	clk_ref = stm32mp_rcc_clock_id_to_clk(clock->clock_id);
@@ -578,14 +656,11 @@ int32_t plat_scmi_rd_autonomous(unsigned int channel_id, unsigned int scmi_id,
 				uint32_t state)
 {
 	const struct stm32_scmi_rd *rd = find_rd(channel_id, scmi_id);
-	unsigned int id = 0;
 
 	if (!rd)
 		return SCMI_NOT_FOUND;
 
-	id = rd->fw_id;
-
-	if (id != STM32MP1_ETZPC_MAX_ID && stm32_etzpc_check_ns_access(id))
+	if (!nsec_can_access_resource(rd->etzpc_id))
 		return SCMI_DENIED;
 	assert(rd->rstctrl);
 
@@ -620,14 +695,11 @@ int32_t plat_scmi_rd_set_state(unsigned int channel_id, unsigned int scmi_id,
 {
 	const struct stm32_scmi_rd *rd = find_rd(channel_id, scmi_id);
 	TEE_Result res = TEE_ERROR_GENERIC;
-	unsigned int id = 0;
 
 	if (!rd)
 		return SCMI_NOT_FOUND;
 
-	id = rd->fw_id;
-
-	if (id != STM32MP1_ETZPC_MAX_ID && stm32_etzpc_check_ns_access(id))
+	if (!nsec_can_access_resource(rd->etzpc_id))
 		return SCMI_DENIED;
 
 #ifdef CFG_STM32MP15
@@ -829,7 +901,7 @@ static TEE_Result stm32mp1_init_scmi_server(void)
 
 			/* Sync SCMI clocks with their targeted initial state */
 			if (clk->enabled &&
-			    stm32mp_nsec_can_access_clock(clk->clock_id))
+			    nsec_can_access_resource(clk->etzpc_id))
 				clk_enable(clk->clk);
 		}
 
-- 
2.47.2

